File: Polars.CSharp.Tests/UdfTests.cs
using Apache.Arrow;
using Polars.NET.Core;
using Pl = Polars.CSharp.Polars; 

namespace Polars.CSharp.Tests;

public static class UdfLogic
{
    public static IArrowArray IntToString(IArrowArray arr)
    {
        if (arr is Int64Array i64Arr)
        {
            var builder = new StringViewArray.Builder();
            for (int i = 0; i < i64Arr.Length; i++)
            {
                if (i64Arr.IsNull(i))
                {
                    builder.AppendNull();
                }
                else
                {
                    long? v = i64Arr.GetValue(i);
                    builder.Append($"Value: {v}");
                }
            }
            return builder.Build();
        }
        
        if (arr is Int32Array i32Arr)
        {
            var builder = new StringViewArray.Builder();
            for (int i = 0; i < i32Arr.Length; i++)
            {
                if (i32Arr.IsNull(i))
                {
                    builder.AppendNull();
                }
                else
                {
                    int? v = i32Arr.GetValue(i);
                    builder.Append($"Value: {v}");
                }
            }
            return builder.Build();
        }

        throw new ArgumentException($"Expected Int32Array or Int64Array, but got: {arr.GetType().Name}");
    }

    public static IArrowArray AlwaysFail(IArrowArray arr)
    {
        throw new Exception("Boom! C# UDF Exploded!");
    }
    public static IArrowArray IntToDouble(IArrowArray arr)
    {
        if (arr is Int64Array i64Arr)
        {
            var builder = new DoubleArray.Builder();
            for (int i = 0; i < i64Arr.Length; i++)
            {
                if (i64Arr.IsNull(i)) 
                {
                    builder.AppendNull();
                }
                else 
                {
                    double val = i64Arr.GetValue(i).Value;
                    builder.Append(val / 2.0);
                }
            }
            return builder.Build();
        }
        if (arr is Int32Array i32Arr)
        {
            var builder = new DoubleArray.Builder();
            for (int i = 0; i < i32Arr.Length; i++)
            {
                if (i32Arr.IsNull(i)) builder.AppendNull();
                else builder.Append(i32Arr.GetValue(i).Value / 2.0);
            }
            return builder.Build();
        }

        throw new ArgumentException($"Expected Int32/64, got {arr.GetType().Name}");
    }
}

public class UdfTests
{
    [Fact]
    public void Map_UDF_Memory_Data_Test()
    {
        int rowCount = 5;
        var builder = new Int64Array.Builder();
        for (int i = 0; i < rowCount; i++) builder.Append(i * 10); // 0, 10, 20, 30, 40
        var arrowArray = builder.Build();

        using var df = DataFrame.FromArrow(
            new RecordBatch.Builder()
                .Append("num", false, col => col.Int64(arr => arr.AppendRange(Enumerable.Range(0, rowCount).Select(x => (long)x * 10))))
                .Build()
        );

        Assert.Equal(5, df.Height);

        Func<IArrowArray, IArrowArray> udf = UdfLogic.IntToDouble;

        using var res = df.Select(
            Pl.Col("num").Map(udf, Pl.Float64).Alias("res")
        );

        Assert.Equal(5, res.Height);

        Assert.NotNull(res.Column("res"));
        Assert.Equal(5, res.Column("res").Length);
        Assert.Equal(0.0, res.Column("res").GetValue<double>(0)); // 0 / 2
        Assert.Equal(20.0, res.Column("res").GetValue<double>(4)); // 40 / 2
    }
    [Fact]
    public void Test_UDF_Map_Stable()
    {
        using var csv = new DisposableFile("num\n15\n25\n" +
                                          "35\n45\n55\n",".csv");
        using var df = DataFrame.ReadCsv(csv.Path);
        Assert.Equal(5, df.Height); 

        // UDF (Int64 -> Int64)
        var udf = Pl.Col("num").Map<long, long>(x => x * 2, typeof(long)).Alias("res");

        using var res = df.Select(
            Pl.Col("num"),
            udf
        );
        Assert.Equal(5, res.Height); 
        
        Assert.Equal(30, res.Column("res").GetValue<long>(0));
        Assert.Equal(50, res.Column("res").GetValue<long>(1));
        Assert.Equal(70, res.Column("res").GetValue<long>(2));
    }
    [Fact]
    public void Map_UDF_Can_Change_Data_Type_Int_To_String()
    {
        using var csv = new DisposableFile("num\n100\n200\n",".csv");
        using var lf = LazyFrame.ScanCsv(csv.Path);

        Func<IArrowArray, IArrowArray> udf = UdfLogic.IntToString;

        using var df = lf.Select(
            Pl.Col("num")
            .Map(udf, DataType.String) 
            .Alias("desc")
        ).Collect();
        df.Show();
        
        Assert.NotNull(df.Column("desc"));
        Assert.Equal("Value: 100", df.Column("desc").GetValue<string>(0)); 
        Assert.Equal("Value: 200", df.Column("desc").GetValue<string>(1));
    }

    [Fact]
    public void Map_UDF_Error_Is_Propagated_To_CSharp()
    {
        using var csv = new DisposableFile("num\n1",".csv");
        using var lf = LazyFrame.ScanCsv(csv.Path);

        Func<IArrowArray, IArrowArray> udf = UdfLogic.AlwaysFail;

        var ex = Assert.Throws<PolarsException>(() => 
        {
            lf.Select(
                Pl.Col("num").Map(udf, Pl.SameAsInput)
            ).Collect();
        });

        Assert.Contains("Boom! C# UDF Exploded!", ex.Message);
    }
    [Fact]
    public void Test_UDF_HighLevel_Numeric()
    {
        using var csv = new DisposableFile("num\n10\n20\n30\n",".csv");
        using var df = DataFrame.ReadCsv(csv.Path);

        var doubleExpr = Pl.Col("num")
            .Map<long, long>(x => x * 2, Pl.Int64)
            .Alias("doubled");

        using var res = df.Select(Pl.Col("num"), doubleExpr);
        
        Assert.Equal(20, res.Column("doubled").GetValue<long>(0)); // 10 * 2
        Assert.Equal(60, res.Column("doubled").GetValue<long>(2)); // 30 * 2
    }

    [Fact]
    public void Test_UDF_HighLevel_String_Manipulation()
    {
        using var csv = new DisposableFile("name\nAlice\nBob\n",".csv");
        using var df = DataFrame.ReadCsv(csv.Path);

        // UDF: "Hello, {name}!"
        var greetExpr = Pl.Col("name")
            .Map<string, string>(name => $"Hello, {name}!", typeof(string))
            .Alias("greeting");

        using var res = df.Select(Pl.Col("name"), greetExpr);
        
        Assert.Equal("Hello, Alice!", res.Column("greeting").GetValue<string>(0));
        Assert.Equal("Hello, Bob!", res.Column("greeting").GetValue<string>(1));
    }

    [Fact]
    public void Test_UDF_HighLevel_Type_Conversion()
    {
        // UDF Int64 -> String
        
        using var csv = new DisposableFile("id\n1001\n1002\n",".csv");
        using var df = DataFrame.ReadCsv(csv.Path);

        var formatExpr = Pl.Col("id")
            .Map<long, string>(id => $"Order-{id}", typeof(string))
            .Alias("order_id");

        using var res = df.Select(Pl.Col("id"), formatExpr);
        
        Assert.Equal("Order-1001", res.Column("order_id").GetValue<string>(0));
        Assert.Equal("Order-1002", res.Column("order_id").GetValue<string>(1));
    }
    [Fact]
    public void Test_UDF_Nullable_Output()
    {
        using var csv = new DisposableFile("num\n10\n0\n20\n",".csv");
        using var df = DataFrame.ReadCsv(csv.Path);

        // 0 -> null (C# null)
        var cleanExpr = Pl.Col("num")
            .Map<long, long?>(x => x == 0 ? null : x, Pl.Int64)
            .Alias("cleaned");

        using var res = df.Select(Pl.Col("num"), cleanExpr);
        
        Assert.Equal(10, res.Column("cleaned").GetValue<long>(0));
        Assert.Null(res.Column("cleaned").GetValue<long?>(1)); // 0 -> Null
        Assert.Equal(20, res.Column("cleaned").GetValue<long>(2));
    }
    [Fact]
    public void Test_UDF_Nullable_Input()
    {
        using var csv = new DisposableFile("num\n10\n\n",".csv"); 
        using var df = DataFrame.ReadCsv(csv.Path);

        // int? -> string
        // null -> "FoundNull" or "Value:{x}"
        var checkNullExpr = Pl.Col("num")
            .Map<long?, string>(x => x.HasValue ? $"Value:{x}" : "FoundNull", Pl.Utf8)
            .Alias("status");

        using var res = df.Select(Pl.Col("num"), checkNullExpr);
        
        Assert.Equal("Value:10", res.Column("status").GetValue<string>(0));
        Assert.Equal("FoundNull", res.Column("status").GetValue<string>(1)); 
    }
    [Fact]
    public void Test_GroupBy_Agg_With_HighLevel_Lambda()
    {
        using var df = DataFrame.FromSeries(
            new Series("key", ["A", "A", "B", "B"]),
            new Series("val", [1L, 2L, 3L, 4L]) 
        );

        static long myGroupLogic(long[] nums)
        {
            if (nums == null || nums.Length == 0) return 0;
            return nums.Max() + 10;
        }

        // GroupBy Agg
        var res = df.GroupBy("key")
                    .Agg(
                        Pl.Col("val")
                        .Implode() 
                        .Map((Func<long[], long>)myGroupLogic, typeof(long)) 
                        .Alias("custom_agg")
                    )
                    .Sort("key");

        // A: [1, 2] -> Max 2 -> +10 = 12
        Assert.Equal(12, res["custom_agg"].GetValue<long>(0));
        
        // B: [3, 4] -> Max 4 -> +10 = 14
        Assert.Equal(14, res["custom_agg"].GetValue<long>(1));
    }
    [Fact]
    [Trait("UDF", "MapMany")]
    public void Test_ExprMap_MultiColumn_Add_ArrowRaw()
    {
        // Arrange
        using var df = Pl.CreateDataFrame(
            ("a", new int[] { 1, 2, 3, 4, 5 }),
            ("b", new int[] { 10, 20, 30, 40, 50 })
        );

        // Act
        using var result = df.Select(
            Pl.Col("a").Map(
                function: arrays => {
                    // arrays[0] = col a, arrays[1] = col b
                    var a = (Int32Array)arrays[0];
                    var b = (Int32Array)arrays[1];
                    var builder = new Int32Array.Builder();
                    for (int i = 0; i < a.Length; i++)
                    {
                        if (a.IsNull(i) || b.IsNull(i))
                            builder.AppendNull();
                        else
                            builder.Append(a.GetValue(i)!.Value + b.GetValue(i)!.Value);
                    }
                    return builder.Build();
                },
                outputType: DataType.Int32,
                additionalExprs: Pl.Col("b")
            )
        );

        // Assert
        var summed = result[0].ToArray<int>();
        var expected = new int[] { 11, 22, 33, 44, 55 };
        Assert.Equal(expected, summed);
    }
}

File: Polars.FSharp.Tests/UDFTests.fs
namespace Polars.FSharp.Tests

module UdfLogic =
    open Apache.Arrow

    let intToString (arr: IArrowArray) : IArrowArray =
        match arr with
        | :? Int64Array as i64Arr ->
            let builder = new StringViewArray.Builder()
            for i in 0 .. i64Arr.Length - 1 do
                if i64Arr.IsNull(i) then builder.AppendNull() |> ignore
                else
                    let v = i64Arr.GetValue(i).Value
                    builder.Append $"Value: {v}" |> ignore
            builder.Build() :> IArrowArray

        | :? Int32Array as i32Arr ->
            let builder = new StringViewArray.Builder()
            for i in 0 .. i32Arr.Length - 1 do
                if i32Arr.IsNull i then
                    builder.AppendNull() |> ignore
                else
                    let v = i32Arr.GetValue(i).Value
                    builder.Append $"Value: {v}" |> ignore
            builder.Build() :> IArrowArray
        |_ -> failwith $"Expected Int32Array or Int64Array, but got: {arr.GetType().Name}"

    let alwaysFail (arr: IArrowArray) : IArrowArray =
        failwith "Boom! F# UDF Exploded!"

open Xunit
open Polars.FSharp
open Apache.Arrow
open System
open Polars.NET.Core

type ``UDF Tests`` () =

    [<Fact>]
    [<Trait("UDF","Basic")>]
    member _.``Map UDF can change data type (Int -> String)`` () =
        use csv = new TempCsv "num\n100\n200"
        let lf = LazyFrame.ScanCsv csv.Path

        let udf = UdfLogic.intToString

        let df =
            lf
            |> LazyFrame.withColumn (
                pl.col("num").MapArrow udf
                |> pl.alias "desc"
            )
            |> LazyFrame.select [ pl.col "desc" ]
            |> LazyFrame.collect

        let arrowBatch = df.ToArrow()
        let strCol = arrowBatch.Column "desc" :?> StringViewArray

        Assert.Equal("Value: 100", strCol.GetString 0)
        Assert.Equal("Value: 200", strCol.GetString 1)

    [<Fact>]
    member _.``Map UDF error is propagated to F#`` () =
        use csv = new TempCsv "num\n1"
        let lf = LazyFrame.ScanCsv csv.Path

        let udf = UdfLogic.alwaysFail

        let ex = Assert.Throws<PolarsException>(fun () ->
            lf
            |> LazyFrame.withColumn (
                pl.col("num").MapArrow(udf, DataType.SameAsInput)
            )
            |> LazyFrame.collect
            |> ignore
        )

        Assert.Contains("Boom! F# UDF Exploded!", ex.Message)
        Assert.Contains("C# UDF Failed", ex.Message)

    [<Fact>]
    member _.``Generic Map UDF with Lambda (Int -> String)`` () =
        use csv = new TempCsv "num\n100\n"
        let lf = LazyFrame.ScanCsv csv.Path

        let myLogic = fun (x: int) -> sprintf "Num: %d" (x + 1)

        let df =
            lf
            |> LazyFrame.withColumn (
                pl.col("num").Map myLogic
                |> pl.alias "res"
            )
            |> LazyFrame.select [ pl.col "res" ]
            |> LazyFrame.collect

        let arrow = df.ToArrow()
        let col = arrow.Column "res" :?> StringViewArray

        Assert.Equal("Num: 101", col.GetString 0)
        Assert.Equal(1, col.Length)
    [<Fact>]
    member _.``UDF: Map with Option (Null Handling)`` () =
        // [10, 20, null]
        use csv = new TempCsv "val\n10\n20\n"
        let lf = LazyFrame.ScanCsv csv.Path

        let logic (opt: int option) =
            match opt with
            | Some x when x > 15 -> Some (x * 2)
            | _ -> None

        let df =
            lf
            |> LazyFrame.withColumn (
                pl.col("val").MapOption logic
                |> pl.alias "res"
            )
            |> LazyFrame.collect

        let arrow = df.ToArrow()
        let col = arrow.Column "res" :?> Int32Array

        // Row 0: 10 -> (<=15) -> None
        Assert.True(col.IsNull 0)

        // Row 1: 20 -> (>15) -> 40
        Assert.Equal(40, col.GetValue(1).Value)

        // Row 2: null -> (None) -> None
        Assert.True(col.IsNull 2)
    [<Fact>]
    [<Trait("UDF","SeriesValueOption")>]
    member _.``Series UDF: MapValueOption (Score to Risk Grade)`` () =

        let data = [
            {| Score = ValueSome 95 |}
            {| Score = ValueSome 75 |}
            {| Score = ValueSome 30 |}
            {| Score = ValueSome 999 |}
            {| Score = ValueNone |}
        ]

        use df = DataFrame.ofRecords data

        let scoreSeries = df.Column "Score"

        let calculateGrade (opt: int voption): string voption =
            match opt with
            | ValueSome s when s >= 90 && s <= 100 -> ValueSome "S"
            | ValueSome s when s >= 70 && s < 90   -> ValueSome "A"
            | ValueSome s when s >= 0  && s < 70   -> ValueSome "B"
            | _ -> ValueNone

        use gradeSeries = scoreSeries |> Series.mapValueOption calculateGrade

        // MapValueOption(calculateGrade, DataType.String)

        use resultDf = df.WithColumns(pl.litSeries(gradeSeries).Alias "Grade")

        Assert.Equal(5L, resultDf.Height)

        Assert.Equal("S", resultDf.Cell<string>("Grade", 0))
        Assert.Equal("A", resultDf.Cell<string>("Grade", 1))
        Assert.Equal("B", resultDf.Cell<string>("Grade", 2))

        Assert.Null(resultDf.Cell<string>("Grade", 3))

        Assert.Null(resultDf.Cell<string>("Grade", 4))
    [<Fact>]
    member _.``UDF: Decimal Map (Series Native)`` () =
        // String ["10.50", "20.25", null]
        let data = ["10.50"; "20.25"; null]
        let s = Series.create("str_vals", data)

        // Cast to Decimal(10, 2)
        let sDec = s |> Series.cast(DataType.Decimal(10, 2))

        let logic (opt: decimal option) =
            opt |> Option.map (fun d -> d * 2m)

        let res = sDec |> Series.mapOption logic

        // 10.50 * 2 = 21.00
        Assert.Equal(21.00m, res.GetValue<decimal> 0)

        // 20.25 * 2 = 40.50
        Assert.Equal(40.50m, res.GetValue<decimal> 1)

        // null -> null
        Assert.True(res.GetValue<decimal option>(2).IsNone)
    [<Fact>]
    member _.``Series: Map (Basic UDF)`` () =
        // Data: [1, 2, 3]
        let s = Series.create("nums", [1; 2; 3])

        // Apply to Series
        let sRes = s |> Series.map (fun x -> x * 10)

        // Verify
        Assert.Equal(10, sRes.GetValue<int> 0)
        Assert.Equal(30, sRes.GetValue<int> 2)

    [<Fact>]
    [<Trait("UDF", "Option")>]
    member _.``Series: Map (Option Handling)`` () =
        // Data: [10, null, 30]
        let s = Series.create("vals", [Some 10; None; Some 30])

        // Logic: Some(x) -> Some(x + 1), None -> Some(-1)
        let logic (opt: int option) =
            match opt with
            | Some x -> Some (x + 1)
            | None -> Some -1

        // Apply
        let sRes = s |> Series.mapOption logic

        // Verify
        Assert.Equal(11, sRes.GetValue<int> 0)
        Assert.Equal(-1, sRes.GetValue<int> 1) // None became -1

    [<Fact>]
    member _.``Series: Map (F# Lambda Sugar)`` () =
        // Data: ["a", "b"]
        let s = Series.create("txt", ["a"; "b"])

        let sRes = s |> Series.map (fun x -> x + "_suffix")

        Assert.Equal("a_suffix", sRes.GetValue<string> 0)
    [<Fact>]
    [<Trait("UDF","ValueOption")>]
    member _.``UDF: Map with ValueOption (String Parsing to Int)`` () =
        let data = [
            {| Code = ValueSome "EMP-1024" |}
            {| Code = ValueSome "EMP-0042" |}
            {| Code = ValueSome "ADMIN-1" |}
            {| Code = ValueSome "EMP-ERR" |}
            {| Code = ValueNone |}
        ]

        let lf = DataFrame.ofRecords(data).Lazy()

        //  string voption -> int voption
        let parseEmpId (opt: string voption) =
            match opt with
            | ValueSome s when s.StartsWith "EMP-" ->
                match Int32.TryParse(s.Substring 4) with
                | true, num -> ValueSome num
                | _ -> ValueNone
            | _ -> ValueNone

        let df =
            lf
            |> LazyFrame.withColumn (
                pl.col "Code"
                |> fun e -> e.MapValueOption parseEmpId
                |> pl.alias "EmpId"
            )
            |> LazyFrame.collect
        // shape: (5, 2)
        // ┌──────────┬───────┐
        // │ Code     ┆ EmpId │
        // │ ---      ┆ ---   │
        // │ str      ┆ i32   │
        // ╞══════════╪═══════╡
        // │ EMP-1024 ┆ 1024  │
        // │ EMP-0042 ┆ 42    │
        // │ ADMIN-1  ┆ null  │
        // │ EMP-ERR  ┆ null  │
        // │ null     ┆ null  │
        // └──────────┴───────┘
        Assert.Equal(5L, df.Height)

        // Row 0: "EMP-1024" -> 1024
        Assert.Equal(1024, df.Cell<int>("EmpId", 0))

        // Row 1: "EMP-0042" -> 42
        Assert.Equal(42, df.Cell<int>("EmpId", 1))

        // Row 2: "ADMIN-1" -> ValueNone -> null
        Assert.False(df.Cell<Nullable<int>>("EmpId", 2).HasValue)

        // Row 3: "EMP-ERR" -> TryParse Fail -> ValueNone -> null
        Assert.False(df.Cell<Nullable<int>>("EmpId", 3).HasValue)

        // Row 4: null -> ValueNone -> null
        Assert.False(df.Cell<Nullable<int>>("EmpId", 4).HasValue)

    [<Fact>]
    [<Trait("UDF","Mapi")>]
    member _.``Series.mapi transforms elements using index and value`` () =
        // Arrange
        let s = pl.series "data" [| 10; 20; 30; 40 |]

        // Act: element + index (10+0, 20+1, 30+2, 40+3)
        let mapped = s |> Series.mapi (fun i v -> v + i)

        // Assert
        Assert.Equal(4L, mapped.Length)
        Assert.Equal<int>(10, mapped.GetValue<int>(0))
        Assert.Equal<int>(21, mapped.GetValue<int>(1))
        Assert.Equal<int>(32, mapped.GetValue<int>(2))
        Assert.Equal<int>(43, mapped.GetValue<int>(3))

    [<Fact>]
    [<Trait("UDF","Mapi")>]
    member _.``Series.mapi allows changing data type based on index`` () =
        // Arrange: int series to formatted string labels
        let s = pl.series "items" [| 100; 200; 300 |]

        // Act
        let labels = s |> Series.mapi (fun i v -> sprintf "#%d: %d" (i + 1) v)

        // Assert
        Assert.Equal(3L, labels.Length)
        Assert.Equal<string>("#1: 100", labels.GetValue<string>(0))
        Assert.Equal<string>("#2: 200", labels.GetValue<string>(1))
        Assert.Equal<string>("#3: 300", labels.GetValue<string>(2))
    [<Fact>]
    [<Trait("UDF", "Mapi")>]
    member _.``Series.mapiOption correctly processes and emits null values`` () =
        // Arrange: [Some 10; None; Some 30]
        let s = pl.series "input" [| Some 10; None; Some 30 |]

        // Act:
        // Index 0: 10 + 0 = 10 -> "Item_0_10"
        // Index 1: None -> None (Preserve null)
        // Index 2: 30 + 2 = 32 -> "Item_2_32"
        let mapped =
            s |> Series.mapiOption (fun i opt ->
                match opt with
                | Some v -> Some (sprintf "Item_%d_%d" i (v + i))
                | None -> None
            )

        // Assert
        Assert.Equal(3L, mapped.Length)
        Assert.False(mapped.IsNullAt 0L)
        Assert.Equal<string>("Item_0_10", mapped.GetValue<string>(0L))

        Assert.True(mapped.IsNullAt 1L)

        Assert.False(mapped.IsNullAt 2L)
        Assert.Equal<string>("Item_2_32", mapped.GetValue<string>(2L))

    [<Fact>]
    [<Trait("UDF", "Mapi")>]
    member _.``Series.mapiValueOption correctly handles ValueOption with zero heap overhead`` () =
        // Arrange: [Some 100.0; None; Some 300.0]
        let s = pl.series "weights" [| ValueSome 100.0; ValueNone; ValueSome 300.0 |]

        // Act: Fill missing with index default, scale existing by index
        let result =
            s |> Series.mapiValueOption (fun i vopt ->
                match vopt with
                | ValueSome v -> ValueSome (v * float (i + 1))
                | ValueNone -> ValueSome (float (i * 999)) // Impute on the fly
            )

        // Assert
        Assert.Equal(3L, result.Length)
        Assert.False(result.IsNullAt 0L)
        Assert.Equal(100.0, result.GetValue<double>(0L))  // 100.0 * 1

        Assert.False(result.IsNullAt 1L)
        Assert.Equal(999.0, result.GetValue<double>(1L))  // Imputed: 1 * 999

        Assert.False(result.IsNullAt 2L)
        Assert.Equal(900.0, result.GetValue<double>(2L))  // 300.0 * 3
    [<Fact>]
    [<Trait("UDF", "Map2")>]
    member _.``Series.map2 element-wise combines two Series matching Array.map2 behavior`` () =
        // Arrange
        let s1 = pl.series "prices" [| 100.0; 200.0; 300.0 |]
        let s2 = pl.series "discounts" [| 0.1; 0.2; 0.05 |]

        // Act: (s1, s2) ||> Series.map2 f
        let netPrices =
            (s1, s2)
            ||> Series.map2 (fun price discount -> price * (1.0 - discount))

        // Assert
        Assert.Equal(3L, netPrices.Length)
        Assert.Equal(90.0, netPrices.GetValue<double>(0))
        Assert.Equal(160.0, netPrices.GetValue<double>(1))
        Assert.Equal(285.0, netPrices.GetValue<double>(2))

    [<Fact>]
    [<Trait("UDF", "Map2")>]
    member _.``Series.mapi2 includes 0-based row index during pairwise mapping`` () =
        // Arrange
        let s1 = pl.series "prefix" [| "Item"; "Item"; "Item" |]
        let s2 = pl.series "sku" [| 101; 202; 303 |]

        // Act: format like "0: Item_101"
        let formatted =
            (s1, s2)
            ||> Series.mapi2 (fun i pref sku -> sprintf "%d: %s_%d" i pref sku)

        // Assert
        Assert.Equal(3L, formatted.Length)
        Assert.Equal<string>("0: Item_101", formatted.GetValue<string>(0))
        Assert.Equal<string>("1: Item_202", formatted.GetValue<string>(1))
        Assert.Equal<string>("2: Item_303", formatted.GetValue<string>(2))

    [<Fact>]
    [<Trait("UDF", "Map2")>]
    member _.``Series.map2 throws ArgumentException when lengths mismatch`` () =
        // Arrange
        let s1 = pl.series "short" [| 1; 2 |]
        let s2 = pl.series "long" [| 1; 2; 3 |]

        // Act & Assert
        Assert.Throws<ArgumentException>(fun () ->
            (s1, s2) ||> Series.map2 (+) |> ignore
        ) |> ignore
